R/doubletFinder.R

```diff
@@ -5,59 +5,69 @@
 #' is performed on the merged real-artificial dataset using PCA. Third, the proportion of
 #' artificial nearest neighbors is defined for each real cell. Finally, real cells are rank-
 #' ordered and predicted doublets are defined via thresholding based on the expected number
-#' of doublets.
+#' of doublets. This function has methods for both Seurat and loom objects. To use the method for loom,
+#' please install loomR (from GitHub, not on CRAN yet) and the loom branch of Seurat. See
+#' https://satijalab.org/seurat/mca_loom.html for more instructions.
 #'
-#' @param seu A fully-processed Seurat object (i.e. after normalization, variable gene definition,
+#' @param obj A fully-processed Seurat or loom object (i.e. after normalization, variable gene definition,
 #' scaling, PCA, and tSNE).
+#' @rdname doubletFinder
```

Comment:
I do not have much experience with object-oriented programming -- Would you mind briefly explaining the purpose of this section?

Reply:
In R, there are a few different systems of object oriented programming, such as S3, S4, and R6. Seurat uses S4, and loomR uses R6. Base R uses S3 a lot. In R, at least for S3 and S4, unlike in Python and C++, methods belong to functions rather than classes. In S3, when a function has methods for multiple classes, you define a generic first, which is what I did in this section. Then the methods for each class are defined as separate functions. How does R know which method to use for an object? For S3, the class for which a function is a method for is specified by `fun.class`, and that's why there're functions `doubletFinder.seurat` for the class Seurat and `doubletFinder.loom` for the class loom. This is also why probably you have seen things like `plot.lm`, since it's a method of function `plot` for the class `lm`. See https://adv-r.hadley.nz/oo.html for a more detailed explanation of object oriented programming in R.